### PR TITLE
Work performed on various issues

### PR DESCRIPTION
**Pull request by ioannis-e**
The highlights of the changes made are as follows:
- Install a hook on LdrpCallInitRoutine() or the relevant nt loader code to cover all Windows XP to Windows 10, x86 and x64 platforms in order to Refresh Modules before the dll entry point. The current implementation Refreshes the Modules after the dll entry point which in effect does not detect any allocations made on dll entry point.
- Fix #9519, #9859, #10544, use LoaderLock to serialize any calls which load additional libraries or require access to the Loader Lock.
- Fix #6359, #10553, fix crashes and failure to register COM dlls where Visual Leak Detector is unloaded before IMalloc interface is released.
- Optimize _GetProcAddress  and _GetProcAddressForCaller to properly test for ordinal vs function name
- Append Visual Studio 2015/2013/2012/2010/2008 symbols cache directory where multiple VS versions are installed on the same system.
- Add option to skip reporting crt startup allocations as memory leaks
 - for debug runtime (where debug crt allocations are not properly identied) and
 - for release runtime (by excluding certain startup crt initialisation modules), ~~this may have an effect in performance.~~
- Improve the vld.ini searching from additional locations listed below in order, and fix installer issue where it would not set the vld.ini path in registry for x64 platform:
 - current directory
 - directory where VLDDLL resides
 - directory where executable resides
 - path specified in registry HKCU
 - path specified in registry HKLM
- All other commits resolve compiler warnings and improvements to the code following PVS-Studio analysis.
- Clean up of solutions and project files
- Improvements and additions to test cases

I tried to keep commits independed of each other as much as possible, to make review easier.

I would suggest to cherypick as much as possible into master in order to rebase this PR and leave us with only the commits that need discussion.

### CODE REVIEW

**Comment by ioannis-e** [comment]
Do you want any changes made from this point on in one or more new commits or edit the relevant commits in place? What's best for review?

**Comment by KindDragon** [comment]
You can make changes in new commit and then rebase before we will merge this PR.

**Comment by ioannis-e** [comment]
Super, we lost the entire review because i had to rebase master due to 778a1e95803b30a3e0918348504455c7343a0cdb

In any case I have resolved the performance degradation issue.
So you are good to go again.

**Comment by ioannis-e** [comment]
I'll have time this weekend to resolve any new review points so we can finally finish this.
Only things left for me to do is:
- [ ] merge review commits with original commits once you give me the ok
- [ ] run a 4 spaces formatting at the end, to have consistent code format.

File: src/tests/vld_main/vld_main.cpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+void* g_m = malloc(30);
+char* g_n = new char[40];
+
+
+int _tmain(int argc, _TCHAR* argv[])
+{
+    void* m = malloc(50);
+    char* n = new char[60];
+
+    return VLDGetLeaksCount() == 8 ? 0 : 1;
```
Comment (KindDragon) [inline src/tests/vld_main/vld_main.cpp]: VLDGetLeaksCount() return 11 for me

Comment (ioannis-e) [inline src/tests/vld_main/vld_main.cpp]: VS2015 ? Debug / Release ?
AppVeyor reports 8 leaks in the Release test for both Win32 and x64.
Can you post the full report so we can check if they are crt leaks

**Comment by KindDragon** [comment]
Your code crash on x64 (Windows 10) when call `FreeLibrary`
Exception thrown at 0x00007FF8870B3D6D (ntdll.dll) in vld_main.exe: 0xC0000005: Access violation reading location 0xFFFFFFFFFFFFFFFF.

File: src/tests/vld_main/vld_main.cpp (hunk 1 above)
Comment (KindDragon) [inline src/tests/vld_main/vld_main.cpp]: I am calling VLDReportLeaks() before return from main
VS2015 Debug return 9 https://gist.github.com/KindDragon/fcb0757383efd6b4018a
VS2015 Debug_StaticCrt return 12 https://gist.github.com/KindDragon/06b1dc5e13f62956eb0c

**Comment by KindDragon** [comment]
Please cheery pick this commit https://github.com/KindDragon/vld/commit/4a91bc753a1b0aea41f5e6a99d3e598dd453edab

**Comment by KindDragon** [comment]
and ComTest doesn't register on x64

**Comment by ioannis-e** [comment]
I have fixed
- crash on x64 (Windows 10) when call ``FreeLibrary``
- ComTest doesn't register on x64
- vld_main additional CRT leaks on VS2015
- improvements to vld_unload test

**Comment by ioannis-e** [comment]
With the modification made to the project files, the vs14 solution is loadable in vs12, the only thing is that the platform should be changed within VS, I have made the platform a default setting for all configurations, so when it changes it changes for all configurations.

In this respect i guess we dont need to keep vs12 and vs14 solutions.
Would you agree to remove the vs12 solutions and projects ?

**Comment by KindDragon** [comment]
Great work

> Would you agree to remove the vs12 solutions and projects ?

I'm using VS12 solution to check vld with VS2013 toolset on local Teamcity. Therefore, it is still necessary to me.

**Comment by ioannis-e** [comment]
Ok, i'll keep them and i will make the solutions and projects the same appart from the toolset.

**Comment by ioannis-e** [comment]
Updated projects and solutions
Added PerUserREdirection to ComTest
Added StaticCrt to ComTest (lets see if anything else pops up on appveyor)

If there is nothing else, i am ready !

**Comment by ioannis-e** [comment]
[I messed up smt... click to view appveyor](https://ci.appveyor.com/project/ioannis-e/vld/build/20)
vldmfc cannot find vld.lib ?
Do you know why ?

**Comment by ioannis-e** [comment]
Builds successfully again!
So let me know how to proceed.

**Comment by KindDragon** [comment]
So everything is ok. If you have already done, I merge PR.

**Comment by ioannis-e** [comment]
Do you want me to squash any commits ? or leave them as they are ?
I'll make another PR for source formatting

**Comment by ioannis-e** [comment]
The following should be resolved:
https://vld.codeplex.com/workitem/10548
https://vld.codeplex.com/workitem/10588
https://vld.codeplex.com/workitem/10587

**Comment by KindDragon** [comment]
If not hard please merge commits please squash commits that fix other commits like 53277cb with b491ca2be, 69f8936 with b38b1e0a88a7e, 3a8a03b53ff99dac with de78327c66aa190cd
Please also add comment to de78327c66aa190cd why we should skip count CRT_USE_TYPE(crtheader->use) == CRT_USE_IGNORE

**Comment by ioannis-e** [comment]
OK ready to merge PR.

On a side note ``safe`` is not that much slower from ``fast``
```
Environment: VldStackWalkMethod=safe; Configuration: Release; Platform: Win32  1 min 35 sec
Environment: VldStackWalkMethod=safe; Configuration: Release; Platform: x64    1 min 39 sec
Environment: VldStackWalkMethod=fast; Configuration: Release; Platform: Win32  1 min 24 sec
Environment: VldStackWalkMethod=fast; Configuration: Release; Platform: x64    1 min 22 sec
```

**Comment by KindDragon** [comment]
> Appveyor runs vld_main in tests but it does not take into account the process exit code.
> Can you fix this, so the exit code is taken into account when running the tests ?

I'll try to think of something

**Comment by Trass3r** [comment]
Nice work!

**Comment by ioannis-e** [comment]
Thanks :+1: actually @KindDragon and his integration of AppVeyor helped a lot!!! 
Hopefully now @KindDragon can release a new version soon with a brand new number ``v2.½`` :)

Also please delete any branches with obsolete code out there, just to be safe.

**Comment by KindDragon** [comment]
Hi @ioannis-e,
I fix vld_main test for AppVeyor and it fail on x64 for some configurations? Can you look at it?

https://ci.appveyor.com/project/KindDragon/vld

**Comment by KindDragon** [comment]
I found what cause it. Win64 optimizer inline `dynamic initializer for '*''() call
Debug:

 	vld_x64.dll!VisualLeakDetector::_HeapAlloc(void * heap, unsigned long flags, unsigned __int64 size) Line 1640	C++
 	ucrtbased.dll!heap_alloc_dbg_internal(const unsigned __int64 size, const int block_use, const char * const file_name, const int line_number) Line 359	C++
 	ucrtbased.dll!heap_alloc_dbg(const unsigned __int64 size, const int block_use, const char * const file_name, const int line_number) Line 450	C++
 	ucrtbased.dll!_malloc_dbg(unsigned __int64 size, int block_use, const char * file_name, int line_number) Line 492	C++
 	ucrtbased.dll!malloc(unsigned __int64 size) Line 22	C++
 	vld_x64.dll!VisualLeakDetector::_malloc(void * (unsigned __int64) * pmalloc, context_t & context, bool debugRuntime, unsigned __int64 size) Line 169	C++
 	vld_x64.dll!CrtPatch<130,1>::crtd_malloc(unsigned __int64 size) Line 431	C++
	vld_main.exe!`dynamic initializer for 's_m''() Line 20	C++
 	ucrtbased.dll!_initterm(void (void) * * first, void (void) * * last) Line 22	C++
 	vld_main.exe!__scrt_common_main_seh() Line 232	C++
 	vld_main.exe!__scrt_common_main() Line 309	C++
 	vld_main.exe!wmainCRTStartup() Line 17	C++
 	kernel32.dll!BaseThreadInitThunk()	Unknown
 	ntdll.dll!RtlUserThreadStart()	Unknown

Release:

	vld_x64.dll!VisualLeakDetector::_HeapAlloc(void * heap, unsigned long flags, unsigned __int64 size) Line 1640	C++
 	ucrtbase.dll!malloc()	Unknown
 	vld_x64.dll!CrtPatch<130,0>::crtd_malloc(unsigned __int64 size) Line 430	C++
 	ucrtbase.dll!_initterm()	Unknown
 	vld_main.exe!__scrt_common_main_seh() Line 232	C++
 	kernel32.dll!BaseThreadInitThunk()	Unknown
 	ntdll.dll!RtlUserThreadStart()	Unknown

**Comment by ioannis-e** [comment]
@KindDragon Thanks, i'll have a look as well over the weekend.

I am considering that maybe ``isCrtStartupModule`` is too agressive for VS2015 due to 
``endWith(filename, len, L"\\crts\\ucrt\\src\\appcrt\\startup\\initterm.cpp") ||``

I'll let you know when i run some tests.

By the way does VS2013 pass all tests ?
I think we may need to add ``endWith(filename, len, L"\\crt\\crtw32\\startup\\tidtable.c") ||``

**Comment by KindDragon** [comment]
> By the way does VS2013 pass all tests ?

Yes, except Release_StaticCrt x64

**Comment by KindDragon** [comment]
> I am considering that maybe isCrtStartupModule is too agressive for VS2015 due to
> endWith(filename, len, L"\\crts\\ucrt\\src\\appcrt\\startup\\initterm.cpp") ||

It does not help us, because we skip allocations from ucrtbase.dll (_initterm) which is excluded module

**Comment by KindDragon** [comment]
b8ff3b1915b7eadb6cca8633119ee7ec04d03a9c

**Comment by ioannis-e** [comment]
Actually this is how i would solve this 12cdc5ff036557658428a525ac6a510d6bf63818
As i said we shouldn't modify ``vld_main_test``

``typedef CrtPatch<130>`` to ``typedef CrtPatch<140>`` really did the trick :)

[AppVeyor agrees with me](https://ci.appveyor.com/project/ioannis-e/vld/build/1)

**Comment by ioannis-e** [comment]
I also tested ``runalltests v120`` with 85f60e85cce2740181b5a294cac31cfb7aa1b7f8 and all tests pass for VS 2013

**Comment by KindDragon** [comment]
A lot of issues with VS2013 toolset :-( https://ci.appveyor.com/project/KindDragon/vld

**Comment by ioannis-e** [comment]
I see, could you also try ioannis-e@85f60e85cce2740181b5a294cac31cfb7aa1b7f8  instead of b8ff3b1915b7eadb6cca8633119ee7ec04d03a9c ?

**Comment by KindDragon** [comment]
https://ci.appveyor.com/project/KindDragon/vld/build/94

**Comment by ioannis-e** [comment]
I am testing as well in my repo.
https://ci.appveyor.com/project/ioannis-e/vld/build/3
i'd suggest to use a ``develop`` branch to test stuff to keep master clean?

**Comment by ioannis-e** [comment]
Lets fix VS2013 first and then do VS2012.
Does appveyor have the 2012 toolset ?
In my pc i couldnt completely build due to debug iterator errors :(

**Comment by ioannis-e** [comment]
``Appveyor: error MSB8020: The build tools for v110_xp (Platform Toolset = 'v110_xp') cannot be found.``
Maybe ask support how to install ?

```
Environment: VldStackWalkMethod=safe, Solution=vld_vs12.sln; Configuration: Release_StaticCrt; Platform: Win32
```
Seems to be deadlocking on appveyor but runs ok localy ...

**Comment by KindDragon** [comment]
> Seems to be deadlocking on appveyor but runs ok localy ...

With StackWalkMethod safe?

**Comment by ioannis-e** [comment]
wait, let me test safe locally... and i'll report back

**Comment by ioannis-e** [comment]
Yes just run ``dynamic_app`` test with ``StackWalkMethod=safe`` and it runs ok locally, but deadlocks in appveyor.
Can you run the same test as well ?

**Comment by ioannis-e** [comment]
https://ci.appveyor.com/project/ioannis-e/vld/build/3

So it only fails on vs12   Release_StaticCrt   Win32/x64 in safe mode
So how is the test different between vs14 and vs12 ?

**Comment by KindDragon** [comment]
Hang for me

**Comment by ioannis-e** [comment]
On which windows are you ?

**Comment by ioannis-e** [comment]
Can you add a ``LoaderLock ll;`` in ``SafeCallstack::getStackTrace``

**Comment by KindDragon** [comment]
win 10

**Comment by KindDragon** [comment]
Works for me. Waiting for appveyor https://ci.appveyor.com/project/KindDragon/vld/build/99

**Comment by KindDragon** [comment]
Now Environment: VldStackWalkMethod=safe, Solution=vld_vs12.sln; Configuration: Debug_VldRelease_StaticCrt; Platform: Win32 hang

**Comment by KindDragon** [comment]
LdrLock deadlock with crt heap lock

**Comment by ioannis-e** [comment]
Damn, thread syncronisations don't work... i thought i got them working (back to the drawing board)

**Comment by ioannis-e** [comment]
See if you like ioannis-e@357f91a3afe8e9ae24908647fba6b5544db6db04 and whether it works for you as well ?
https://ci.appveyor.com/project/ioannis-e/vld/build/7

I temporarily removed VS2012 appveyour tests since they kept failing. Can you build/run in VS2012/VS2010 at all ?
I am interested to see whether ``vld_main_test`` and ``vld_unload`` exhibit different behaviour.

**Comment by KindDragon** [comment]
> Can you build/run in VS2012/VS2010 at all ?

I can not configure build under VS2010

**Comment by ioannis-e** [comment]
I meant locally not on appveyor.
I have VS2010 on x86, but the ``Batch Build...`` feature is broken, and i cannot test all configurations.
I want to confirm that ``isCrtStartupModule()`` reports correctly in older versions.

Did you get any feedback from appveyor support regarding v110 toolset ?

**Comment by KindDragon** [comment]
> I meant locally not on appveyor.

I meant this too

**Comment by KindDragon** [comment]
> See if you like ioannis-e@357f91a

I don't know how I fill about this changes :-) Too many critical sections and locks.
Also, help for DbgHelp doesn't guarantee that your changes will work.

> All DbgHelp functions, are single threaded. Therefore, calls from more than one thread to this function will likely result in unexpected behavior or memory corruption. To avoid this, you must synchronize all concurrent calls from more than one thread to this function.
But you use 4 CS.

**Comment by ioannis-e** [comment]
@KindDragon  I understand, so let me explain my reasoning.

There were already 2 locks in code ``g_symbolLock`` and ``g_imageLock`` so I guess there was a reasoning behind having two locks instead of just one lock.

Using ``g_symbolLock``with ``StackWalk64`` caused a deadlock in ``safe`` and ``StaticCrt`` mode.
I added 2 more one for ``StackWalk64`` and one for ``EnumerateLoadedModulesW64``
I kept the same lock for all other ``Sym*`` functions.

Also I used the wrapper functions in order to avoid unecessary blocking or neigbouring code by the locks.

Regarding ``DbgHelp`` help its unclear to me whether **all** functions of ``DbgHelp`` need to be syncronised between threads or whether calls to the **same** function need to be syncronised between threads, I went with the latter and haven't identified any corruption yet...

**Comment by KindDragon** [comment]
I also fix deadlock in Magic commit :-) 9902f53e12ccf761208f78601b42024747da7969
But I do not remember why I added heapMapLock earlier

**Comment by KindDragon** [comment]
heapMapLock was added not sol long ago https://github.com/KindDragon/vld/commit/b9a53cd07818539045248e812dbdc52211f38ce0 instead of g_stackWalkLock

**Comment by KindDragon** [comment]
> Also I used the wrapper functions in order to avoid unecessary blocking or neigbouring code by the locks.

But frequent of the entry/exit the critical section is slower than once to enter before loop.

**Comment by KindDragon** [comment]
How you think if I partially revert b9a53cd ("CS g_stackWalkLock merged with g_heapMapLock").
I think this should fix deadlock

**Comment by ioannis-e** [comment]
You see upto that commit each DbgHelp function had its own locker and all other Sym* functions had one locker, which is basically what i did but instead put them in a class.

So I guess you should go back to g_stackWalkLock for StackWalk64 and probably add one more lock for EnumerateModulesW64.

What I also considered is that StackWalk64 and EnumerateModulesW64 are the **most time consuming functions**

I would also suggest the following shanges
m_modulesLock does not need to block IsModulePatched because that function acuires other locks that could lead to deadlock
https://github.com/ioannis-e/vld/commit/357f91a#diff-a895db1a9b34fcae7efbc9c48edaf1c4L726
and
Also I think nobody likes ``IsBadReadPtr`` :)
https://github.com/ioannis-e/vld/commit/357f91a#diff-a895db1a9b34fcae7efbc9c48edaf1c4L723

**Comment by ioannis-e** [comment]
I finally managed to build under VS2010.

I have an issue with ``_heap_term()`` where in VS2010 it calls ``HeapDestroy`` while in VS2013 and VS2015 it does not.

This causes some ``StaticCrt`` tests ie ``dynamic_app`` and ``vld_unload`` to fail in VS2010 as the leaks are reported before destroying the heap (in order to perform the memory dump) and the heap map is erased, and are not taken into account when calling ``GetLeakCount``.

I was wandering whether you would accept on ``HeapDestroy`` in addition to reporting the leaks immediately, to keep the heap info (if leaks are found) and resolve the stacks in order to be able to report the leaks when vld unloads without a memory dump.

I'm currently working on some changes to see how it goes.

**Comment by KindDragon** [comment]
> Did you get any feedback from appveyor support regarding v110 toolset ?

They use VS Express and it doesn't contain MFC

**Comment by KindDragon** [comment]
> I finally managed to build under VS2010.

VS2012 build failed for Environment: VldStackWalkMethod=fast, Solution=vld_vs11_wo_mfc.sln, GTEST_FILTER=-*.Mfc*; Configuration: Release_StaticCrt; Platform: Win32
https://ci.appveyor.com/project/KindDragon/vld/build/119

**Comment by KindDragon** [comment]
> This causes some StaticCrt tests ie dynamic_app and vld_unload to fail in VS2010 as the leaks are reported before destroying the heap (in order to perform the memory dump) and the heap map is erased, and are not taken into account when calling GetLeakCount.

Maybe just change leaks count for this configuration (will simply be known difference in behavior). Someday we will remove support for VS2010

**Pull request merged**